common/lib/xmodule/xmodule/contentstore/mongo.py

```diff
@@ -247,19 +247,64 @@ def _get_all_content_for_course(self,
             contentType: The mimetype string of the asset
             md5: An md5 hash of the asset content
         '''
-        query = query_for_course(course_key, "asset" if not get_thumbnails else "thumbnail")
-        find_args = {"sort": sort}
-        if maxresults > 0:
-            find_args.update({
-                "skip": start,
-                "limit": maxresults,
-            })
+        # TODO: Using an aggregate() instead of a find() here is a hack to get around the fact that Mongo 3.2 does not
+        # support sorting case-insensitively.
+        # If a sort on displayname is requested, the aggregation pipeline creates a new field:
+        # `insensitive_displayname`, a lowercase version of `displayname` that is sorted on instead.
+        # Mongo 3.4 does not require this hack. When upgraded, change this aggregation back to a find and specifiy
+        # a collation based on user's language locale instead.
+        pipeline_stages = []
+        query = query_for_course(course_key, 'asset' if not get_thumbnails else 'thumbnail')
         if filter_params:
             query.update(filter_params)
+        pipeline_stages.append({'$match': query})
+
+        if sort:
+            sort = dict(sort)
+            if 'displayname' in sort:
+                pipeline_stages.append({
+                    '$project': {
+                        'contentType': 1,
+                        'locked': 1,
+                        'chunkSize': 1,
+                        'content_son': 1,
+                        'displayname': 1,
+                        'filename': 1,
+                        'length': 1,
+                        'import_path': 1,
+                        'uploadDate': 1,
+                        'thumbnail_location': 1,
+                        'md5': 1,
+                        'insensitive_displayname': {
+                            '$toLower': '$displayname'
```

Comment:
How does this handle numbers?

Reply:
I tried this out on the test course on the sandbox: https://studio-thallada.sandbox.edx.org/assets/course-v1:edX+Test101+course/

<img width="606" alt="screen shot 2018-02-02 at 4 21 51 pm" src="https://user-images.githubusercontent.com/9357438/35755446-477f35dc-0835-11e8-9ab9-ae390b13efee.png">

Reply:
There might be a "natural" sort package one could implement or use--not sure how painful it is to get the sorting working in such a way.

Reply:
This doesn't seem like an easy thing to do with mongo. And @dsjen, unfortunately we can't do it in JS because the JS never gets access to the full list of files, just one page.

I think the only way to implement that would be to use [`mapReduce`](https://docs.mongodb.com/manual/reference/command/mapReduce/index.html) to run a Javascript function over each element that finds any numbers at the start of displaystring and emits them as a separate integer field and then have the $sort sort on that field after sorting on displayname.

That seems pretty bad for performance adds enough complexity that I think it would just be simpler for us to switch to elasticsearch...

Reply:
Bummer! Thanks for researching that. I think what you have is miles better than what was in before, and we should go with it.

Reply:
Also, it looks like Mongo 3.4's [collation](https://docs.mongodb.com/manual/reference/collation/) has a `numericOrdering` option that should just do the right thing.